File: src/main/java/ru/arc/autobuild/ConstructionSite.java
package ru.arc.autobuild;

import lombok.ToString;
import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.hooks.HookRegistry;
import ru.arc.util.CooldownManager;
import ru.arc.util.Utils;
import com.sk89q.worldedit.math.BlockVector3;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.text.Component;
import org.bukkit.*;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Firework;
import org.bukkit.entity.Player;
import org.bukkit.inventory.meta.FireworkMeta;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

import static org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CUSTOM;
import static ru.arc.util.Logging.error;
import static ru.arc.util.Logging.info;

@Slf4j
@RequiredArgsConstructor
@Getter
@ToString
public class ConstructionSite {

    final Building building;
    final Location centerBlock;
    final Player player;
    final int rotation;
    final World world;
    final int subRotation;
    final int yOffset;
    State state = State.CREATED;
    Set<Chunk> chunks = new HashSet<>();
    long timestamp = System.currentTimeMillis();

    Display display;
    Construction construction;

    Config config = ConfigManager.of(ARC.plugin.getDataPath(), "auto-build.yml");

    int npcId = -1;

    Color[] colors = new Color[]{Color.WHITE, Color.AQUA, Color.PURPLE, Color.YELLOW, Color.MAROON, Color.GREEN, Color.TEAL,
            Color.OLIVE, Color.FUCHSIA, Color.LIME, Color.RED, Color.ORANGE};

    public void startDisplayingBorder(int seconds) {
        if (display != null) throw new IllegalStateException("Display is not null when trying to build!");

        display = new Display(this);
        display.showBorder(seconds);
        state = State.DISPLAYING_OUTLINE;
    }

    public void startDisplayingBorderAndDisplay(int seconds) {
        if (display == null || state != State.DISPLAYING_OUTLINE)
            throw new IllegalStateException("Display is null or state is not outline!");

        display.showBorderAndDisplay(seconds);
    }

    public Location getCenterBlock() {
        return centerBlock.clone().add(0, yOffset, 0);
    }

    List<Location> getCenterLocations() {
        Location center1 = this.centerBlock.toBlockLocation().clone().add(-0.05, -1.05, -0.05);
        Location center2 = center1.clone().add(1.1, 1.1, 1.1);
        return Utils.getBorderLocations(center1, center2, 6);
    }

    List<Utils.LocationData> getBorderLocations() {
        ConstructionSite.Corners corners = this.getCorners();

        Location corner1 = new Location(this.getWorld(),
                corners.corner1().x() + this.getCenterBlock().x(),
                corners.corner1().y() + this.getCenterBlock().y(),
                corners.corner1().z() + this.getCenterBlock().z());

        Location corner2 = new Location(this.getWorld(),
                corners.corner2().x() + this.getCenterBlock().x() + 1,
                corners.corner2().y() + this.getCenterBlock().y() + 1,
                corners.corner2().z() + this.getCenterBlock().z() + 1);

        return Utils.getBorderLocationsWithCornerData(corner1, corner2, 2, 3);
    }

    public int fullRotation() {
        return (rotation + subRotation) % 360;
    }

    public void spawnConfirmNpc(int seconds) {
        if (construction != null) throw new IllegalStateException("Construction is not null when creating NPC!");
        this.timestamp = System.currentTimeMillis();
        construction = new Construction(this);
        npcId = construction.createNpc(centerBlock, seconds);
        state = State.CONFIRMATION;
    }

    public void startBuild() {
        if (construction == null) throw new IllegalStateException("Construction is null when trying to build!");
        display.stop();
        forceloadChunks();
        construction.startBuilding();
        if (!player.hasPermission("arc.buildings.bypass-cooldown")) {
            CooldownManager.addCooldown(player.getUniqueId(), "building_cooldown", 20 * 60 * 60);
        }
        state = State.BUILDING;
    }

    public void stopBuild() {
        if (construction != null) {
            construction.cancel(0);
            state = State.CANCELLED;

            cleanup(0);
        } else {
            throw new IllegalStateException("Stoping build when not building!");
        }
    }

    public void stopOutlineDisplay() {
        if (display != null && state == State.DISPLAYING_OUTLINE) {
            display.stop();
            state = State.CREATED;
        } else {
            throw new IllegalStateException("Stoping display when not displaying!");
        }
    }

    public void stopConfirmStep() {
        if (display != null && construction != null && state == State.CONFIRMATION) {
            display.stop();
            construction.destroyNpc();
            state = State.CREATED;
        } else {
            throw new IllegalStateException("Stoping confirm when not confirming!");
        }
    }

    public void calculateChunks() {
        if (!chunks.isEmpty()) return;
        Corners corners = getCorners();
        for (int x = corners.corner1.x(); x < corners.corner2.x(); x++) {
            for (int z = corners.corner1.z(); z < corners.corner2.z(); z++) {
                Location location = new Location(world, x + centerBlock.x(), 1, z + centerBlock.z());
                Chunk chunk = location.getChunk();
                chunks.add(chunk);
            }
        }
    }

    public boolean canBuild() {
        calculateChunks();
        if (HookRegistry.landsHook != null) {
            for (Chunk chunk : chunks) {
                if (!HookRegistry.landsHook.canBuild(player, chunk)) {
                    info("Can't build in chunk: {}", chunk);
                    return false;
                }
            }
        }
        if (HookRegistry.wgHook != null) {
            Corners corners = getCorners();
            for (int x = corners.corner1.x(); x < corners.corner2.x(); x++) {
                for (int y = corners.corner1.y(); y < corners.corner2.y(); y++) {
                    for (int z = corners.corner1.z(); z < corners.corner2.z(); z++) {
                        if (!HookRegistry.wgHook.canBuild(player, new Location(world, x, y, z))) {
                            info("Can't build in worldguard: {} {} {}", x, y, z);
                            return false;
                        }
                    }
                }
            }
        }
        return true;
    }

    public void forceloadChunks() {
        calculateChunks();
        info("Forceloading {} chunks", chunks.size());
        for (Chunk chunk : chunks) {
            chunk.setForceLoaded(true);
        }
    }

    public void stopForceload() {
        chunks.stream().filter(Chunk::isForceLoaded).forEach(c -> c.setForceLoaded(false));
    }

    public void finishBuildStateAndCleanup() {
        construction.finishInstantly();
        cleanup(0);
    }

    public double getProgress() {
        if (state != State.BUILDING) return 0;
        int built = construction.pointer.get();
        return ((double) built) / building.volume();
    }


    public record Corners(BlockVector3 corner1, BlockVector3 corner2) {
    }

    public boolean same(Player player, Location location, Building building) {
        return location.toCenterLocation().equals(centerBlock.toCenterLocation()) && building == this.building &&
                rotation == BuildingManager.rotationFromYaw(player.getYaw());
    }


    public Corners getCorners() {
        BlockVector3 c1 = building.getCorner1(fullRotation());
        BlockVector3 c2 = building.getCorner2(fullRotation());

        BlockVector3 r1 = BlockVector3.at(Math.min(c1.x(), c2.x()),
                Math.min(c1.y(), c2.y()),
                Math.min(c1.z(), c2.z()));
        BlockVector3 r2 = BlockVector3.at(Math.max(c1.x(), c2.x()),
                Math.max(c1.y(), c2.y()),
                Math.max(c1.z(), c2.z()));
        r1 = r1.add(0, yOffset, 0);
        r2 = r2.add(0, yOffset, 0);
        return new Corners(r1, r2);
    }


    public void finalizeBuilding() {
        sendFinalMessage();
        launchFireWorks();
        Bukkit.getScheduler().runTaskLater(ARC.plugin, construction::destroyNpc, 60);
        state = State.DONE;

        cleanup(60);
    }

    private void sendFinalMessage() {
        Component message = config.componentDef("building-finished-message", "<gray>\uD83D\uDEE0 <green>Строительство завершено!");
        player.sendMessage(message);
    }

    private void launchFireWorks() {
        try {
            AtomicInteger counter = new AtomicInteger(0);
            new BukkitRunnable() {
                @Override
                public void run() {
                    if (counter.incrementAndGet() >= 5) this.cancel();
                    var velocity = new org.bukkit.util.Vector(0, ThreadLocalRandom.current().nextDouble(0.1, 0.5), 0);
                    Firework firework = (Firework) world.spawnEntity(centerBlock,
                            EntityType.FIREWORK_ROCKET,
                            CUSTOM,
                            e -> e.setVelocity(velocity));
                    FireworkEffect effect = FireworkEffect.builder()
                            .flicker(ThreadLocalRandom.current().nextBoolean())
                            .trail(ThreadLocalRandom.current().nextBoolean())
                            .with(Utils.random(FireworkEffect.Type.values()))
                            .withColor(Utils.random(colors, 3))
                            .withFade(Utils.random(colors, 3))
                            .build();

                    FireworkMeta meta = firework.getFireworkMeta();
                    meta.setPower(ThreadLocalRandom.current().nextInt(1, 3));
                    meta.addEffect(effect);
                    firework.setFireworkMeta(meta);
                    Bukkit.getScheduler().runTaskLater(ARC.plugin, firework::detonate, 20);
                }
            }.runTaskTimer(ARC.plugin, 0L, 10L);
        } catch (Exception e) {
            error("Failed to launch fireworks", e);
        }
    }

    public void cleanup(int destroyNpcDelaySeconds) {
        BuildingManager.removeConstruction(this);
        stopForceload();
        if (display != null) {
            display.stop();
        }
        if (construction != null) construction.cancel(destroyNpcDelaySeconds);
    }

    public enum State {
        CREATED, DISPLAYING_OUTLINE, CONFIRMATION, BUILDING, CANCELLED, DONE
    }

}

File: src/main/kotlin/ru/arc/autobuild/ConstructionSite.kt
package ru.arc.autobuild

import com.sk89q.worldedit.math.BlockVector3
import org.bukkit.Bukkit
import org.bukkit.Chunk
import org.bukkit.Color
import org.bukkit.FireworkEffect
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.entity.EntityType
import org.bukkit.entity.Firework
import org.bukkit.entity.Player
import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CUSTOM
import ru.arc.ARC
import ru.arc.hooks.HookRegistry
import ru.arc.hooks.worldguard.WGHook
import ru.arc.util.LocationUtils
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import ru.arc.util.RandomUtils
import java.util.concurrent.ThreadLocalRandom

/**
 * Represents an active construction site with state management.
 * Uses State Pattern for clean lifecycle management.
 *
 * Lifecycle: Created -> DisplayingOutline -> Confirmation -> Building -> Done
 *                                         \-> Cancelled
 *                            \-> Cancelled
 */
class ConstructionSite(
    val building: Building,
    val centerBlock: Location,
    val player: Player,
    val rotation: Int,
    val world: World,
    val subRotation: Int,
    val yOffset: Int
) {
    /** Corner coordinates for the building area */
    data class Corners(val corner1: BlockVector3, val corner2: BlockVector3)

    // ==================== State Management ====================

    var state: ConstructionState = ConstructionState.Created
        private set

    /**
     * Transitions to a new state with validation.
     * @return true if transition was successful
     */
    fun transitionTo(newState: ConstructionState): Boolean {
        if (!state.canTransitionTo(newState)) {
            error("Invalid transition from {} to {}", state, newState)
            return false
        }

        info("ConstructionSite transition: {} -> {}", state::class.simpleName, newState::class.simpleName)
        state.exit(this)
        state = newState
        state.enter(this)
        return true
    }

    // ==================== Configuration ====================

    /** Seconds to display border before timeout */
    val displaySeconds: Int get() = 180

    /** Seconds to wait for confirmation before timeout */
    val confirmSeconds: Int get() = BuildConfig.confirmTimeSeconds

    // ==================== Mutable Properties ====================

    var timestamp: Long = System.currentTimeMillis()
        internal set

    var npcId: Int = -1
        internal set

    private val chunks = mutableSetOf<Chunk>()
    internal var display: Display? = null
    internal var construction: Construction? = null

    // ==================== Computed Properties ====================

    /** Combined rotation from player direction and sub-rotation */
    val fullRotation: Int get() = (rotation + subRotation) % 360

    /** Center location adjusted for Y offset */
    val adjustedCenter: Location get() = centerBlock.clone().add(0.0, yOffset.toDouble(), 0.0)

    /** Bounding box corners of the construction area */
    val corners: Corners
        get() {
            val c1 = building.getCorner1(fullRotation)
            val c2 = building.getCorner2(fullRotation)
            return Corners(
                BlockVector3.at(minOf(c1.x(), c2.x()), minOf(c1.y(), c2.y()) + yOffset, minOf(c1.z(), c2.z())),
                BlockVector3.at(maxOf(c1.x(), c2.x()), maxOf(c1.y(), c2.y()) + yOffset, maxOf(c1.z(), c2.z()))
            )
        }

    /** Building progress (0.0 to 1.0) */
    val progress: Double
        get() {
            if (state != ConstructionState.Building) return 0.0
            val built = construction?.pointer?.get() ?: return 0.0
            return built.toDouble() / building.volume
        }

    // ==================== Convenience Methods ====================

    /** Start displaying the border outline */
    fun startDisplayingBorder() = transitionTo(ConstructionState.DisplayingOutline)

    /** Move to confirmation phase (spawn NPC) */
    fun startConfirmation() = transitionTo(ConstructionState.Confirmation)

    /** Start building */
    fun startBuild() = transitionTo(ConstructionState.Building)

    /** Cancel construction */
    fun cancel() = transitionTo(ConstructionState.Cancelled)

    /** Mark as done */
    fun complete() = transitionTo(ConstructionState.Done)

    /** Finish building instantly (admin command). Returns true if successful. */
    fun finishInstantly(): Boolean {
        if (state != ConstructionState.Building) return false
        construction?.finishInstantly()
        return transitionTo(ConstructionState.Done)
    }

    // ==================== Permission Checking ====================

    fun canBuild(): Boolean {
        calculateChunks()

        // Check Lands plugin (chunk-based, already efficient)
        HookRegistry.landsHook?.let { lands ->
            for (chunk in chunks) {
                if (!lands.canBuild(player, chunk)) {
                    info("Can't build in chunk: {}", chunk)
                    return false
                }
            }
        }

        // Check WorldGuard - optimized to check only strategic points
        HookRegistry.wgHook?.let { wg ->
            if (!canBuildWorldGuard(wg)) return false
        }
        return true
    }

    /**
     * Optimized WorldGuard check - instead of checking every block (O(n³)),
     * we check corners, edges, and a sample of interior points.
     */
    private fun canBuildWorldGuard(wg: WGHook): Boolean {
        val c = corners
        val checkPoints = mutableSetOf<Triple<Int, Int, Int>>()

        // Add all 8 corners
        for (x in listOf(c.corner1.x(), c.corner2.x())) {
            for (y in listOf(c.corner1.y(), c.corner2.y())) {
                for (z in listOf(c.corner1.z(), c.corner2.z())) {
                    checkPoints.add(Triple(x, y, z))
                }
            }
        }

        // Add edge midpoints
        val midX = (c.corner1.x() + c.corner2.x()) / 2
        val midY = (c.corner1.y() + c.corner2.y()) / 2
        val midZ = (c.corner1.z() + c.corner2.z()) / 2

        for (y in listOf(c.corner1.y(), c.corner2.y())) {
            for (z in listOf(c.corner1.z(), c.corner2.z())) {
                checkPoints.add(Triple(midX, y, z))
            }
        }
        for (x in listOf(c.corner1.x(), c.corner2.x())) {
            for (z in listOf(c.corner1.z(), c.corner2.z())) {
                checkPoints.add(Triple(x, midY, z))
            }
        }
        for (x in listOf(c.corner1.x(), c.corner2.x())) {
            for (y in listOf(c.corner1.y(), c.corner2.y())) {
                checkPoints.add(Triple(x, y, midZ))
            }
        }

        // Add face centers and center point
        checkPoints.add(Triple(midX, midY, c.corner1.z()))
        checkPoints.add(Triple(midX, midY, c.corner2.z()))
        checkPoints.add(Triple(midX, c.corner1.y(), midZ))
        checkPoints.add(Triple(midX, c.corner2.y(), midZ))
        checkPoints.add(Triple(c.corner1.x(), midY, midZ))
        checkPoints.add(Triple(c.corner2.x(), midY, midZ))
        checkPoints.add(Triple(midX, midY, midZ))

        for ((x, y, z) in checkPoints) {
            val loc = Location(world, centerBlock.x + x, centerBlock.y + y, centerBlock.z + z)
            if (!wg.canBuild(player, loc)) {
                info("Can't build in worldguard: {} {} {}", loc.blockX, loc.blockY, loc.blockZ)
                return false
            }
        }

        return true
    }

    // ==================== Chunk Management ====================

    private fun calculateChunks() {
        if (chunks.isNotEmpty()) return
        val c = corners
        for (x in c.corner1.x() until c.corner2.x()) {
            for (z in c.corner1.z() until c.corner2.z()) {
                val loc = Location(world, (x + centerBlock.x), 1.0, (z + centerBlock.z))
                chunks.add(loc.chunk)
            }
        }
    }

    internal fun forceloadChunks() {
        calculateChunks()
        info("Forceloading {} chunks", chunks.size)
        chunks.forEach { it.isForceLoaded = true }
    }

    private fun stopForceload() {
        chunks.filter { it.isForceLoaded }.forEach { it.isForceLoaded = false }
    }

    // ==================== Utilities ====================

    /** Checks if this site matches the given parameters (for click confirmation) */
    fun same(player: Player, location: Location, building: Building): Boolean =
        location.toCenterLocation() == centerBlock.toCenterLocation() &&
            building.fileName == this.building.fileName &&
            rotation == BuildingManager.rotationFromYaw(player.yaw)

    /** Gets particle locations for the center block highlight */
    fun getCenterLocations(): List<Location> {
        val center1 = centerBlock.toBlockLocation().clone().add(-0.05, -1.05, -0.05)
        val center2 = center1.clone().add(1.1, 1.1, 1.1)
        return LocationUtils.getBorderLocations(center1, center2, 6)
    }

    /** Gets particle locations for the building border */
    fun getBorderLocations(): List<LocationUtils.LocationData> {
        val c = corners
        val corner1 = Location(
            world,
            c.corner1.x() + adjustedCenter.x,
            c.corner1.y() + adjustedCenter.y,
            c.corner1.z() + adjustedCenter.z
        )
        val corner2 = Location(
            world,
            c.corner2.x() + adjustedCenter.x + 1,
            c.corner2.y() + adjustedCenter.y + 1,
            c.corner2.z() + adjustedCenter.z + 1
        )
        return LocationUtils.getBorderLocationsWithCornerData(corner1, corner2, 2, 3)
    }

    /** Launches celebratory fireworks */
    internal fun launchFireworks() {
        try {
            val maxFireworks = 5
            var count = 0
            lateinit var task: org.bukkit.scheduler.BukkitTask
            task = Bukkit.getScheduler().runTaskTimer(ARC.plugin, Runnable {
                if (++count > maxFireworks) {
                    task.cancel()
                    return@Runnable
                }
                spawnFirework()
            }, 0L, 10L)
        } catch (e: Exception) {
            error("Failed to launch fireworks", e)
        }
    }

    private fun spawnFirework() {
        val velocity = org.bukkit.util.Vector(0.0, ThreadLocalRandom.current().nextDouble(0.1, 0.5), 0.0)
        val firework = world.spawnEntity(centerBlock, EntityType.FIREWORK_ROCKET, CUSTOM) {
            it.velocity = velocity
        } as Firework

        val colors = listOf(
            Color.WHITE, Color.AQUA, Color.PURPLE, Color.YELLOW, Color.MAROON,
            Color.GREEN, Color.TEAL, Color.OLIVE, Color.FUCHSIA, Color.LIME,
            Color.RED, Color.ORANGE
        )

        val effect = FireworkEffect.builder()
            .flicker(ThreadLocalRandom.current().nextBoolean())
            .trail(ThreadLocalRandom.current().nextBoolean())
            .with(RandomUtils.random(FireworkEffect.Type.entries.toTypedArray()))
            .withColor(RandomUtils.random(colors.toTypedArray(), 3).toList())
            .withFade(RandomUtils.random(colors.toTypedArray(), 3).toList())
            .build()

        firework.fireworkMeta = firework.fireworkMeta.apply {
            power = ThreadLocalRandom.current().nextInt(1, 3)
            addEffect(effect)
        }

        Bukkit.getScheduler().runTaskLater(ARC.plugin, Runnable { firework.detonate() }, 20)
    }

    /** Cleans up all resources */
    internal fun cleanup(destroyNpcDelaySeconds: Int) {
        BuildingManager.removeConstruction(this)
        stopForceload()
        display?.stop()
        construction?.cancel(destroyNpcDelaySeconds)
    }

    override fun toString() =
        "ConstructionSite(building=$building, player=${player.name}, state=${state::class.simpleName})"
}
